core/src/main/java/org/apache/iceberg/AllDataFilesTable.java

```diff
@@ -0,0 +1,149 @@
+    return table.currentSnapshot().manifestListLocation();
+  }
+
+  public static class AllDataFilesTableScan extends BaseTableScan {
+    private static final long TARGET_SPLIT_SIZE = 32 * 1024 * 1024; // 32 MB
+    private final Schema fileSchema;
+
+    AllDataFilesTableScan(TableOperations ops, Table table, Schema fileSchema) {
+      super(ops, table, fileSchema);
+      this.fileSchema = fileSchema;
+    }
+
+    private AllDataFilesTableScan(
+        TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+        boolean caseSensitive, boolean colStats, Collection<String> selectedColumns, Schema fileSchema,
+        ImmutableMap<String, String> options) {
+      super(ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+      this.fileSchema = fileSchema;
+    }
+
+    @Override
+    protected TableScan newRefinedScan(
+        TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+        boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+        ImmutableMap<String, String> options) {
+      return new AllDataFilesTableScan(
+          ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, fileSchema, options);
+    }
+
+    @Override
+    public TableScan useSnapshot(long scanSnapshotId) {
+      throw new UnsupportedOperationException("Cannot select snapshot: all_data_files is for all snapshots");
+    }
+
+    @Override
+    public TableScan asOfTime(long timestampMillis) {
+      throw new UnsupportedOperationException("Cannot select snapshot: all_data_files is for all snapshots");
+    }
+
+    @Override
+    protected long targetSplitSize(TableOperations ops) {
+      return TARGET_SPLIT_SIZE;
+    }
+
+    @Override
+    protected CloseableIterable<FileScanTask> planFiles(
+        TableOperations ops, Snapshot snapshot, Expression rowFilter, boolean caseSensitive, boolean colStats) {
+      CloseableIterable<ManifestFile> manifests = allManifestFiles(ops.current().snapshots());
+      String schemaString = SchemaParser.toJson(schema());
+      String specString = PartitionSpecParser.toJson(PartitionSpec.unpartitioned());
+      ResidualEvaluator residuals = ResidualEvaluator.unpartitioned(rowFilter);
+
+      // Data tasks produce the table schema, not the projection schema and projection is done by processing engines.
+      // This data task needs to use the table schema, which may not include a partition schema to avoid having an
+      // empty struct in the schema for unpartitioned tables. Some engines, like Spark, can't handle empty structs in
+      // all cases.
+      return CloseableIterable.transform(manifests, manifest ->
+          new DataFilesTable.ManifestReadTask(ops.io(), manifest, fileSchema, schemaString, specString, residuals));
+    }
+  }
+
+  static CloseableIterable<ManifestFile> allManifestFiles(List<Snapshot> snapshots) {
```

Comment:
There are a couple of places when we define static methods in one metadata table class and call them in others. It seems we can put some of those in the parent `BaseMetadataTable`.